torch_xla_py/xla_dist.py

```diff
@@ -78,38 +114,157 @@ def validate(self):
+
+  def __init__(self, tpus, vms=None, zone=None, project=None):
     """Creates a new ClusterResolver object."""
 
-    assert tpus, "TPU name list must not be empty."
+    if not isinstance(tpus, list) or len(tpus) == 0:
+      raise ValueError('tpus must be a non-empty list')
+    if vms is not None:
+      if not isinstance(vms, list) or len(vms) == 0:
+        raise ValueError('vms must be a non-empty list if provided')
 
     self._tpus = tpus
     self._vms = vms
     self._zone = zone
     self._project = project
 
+    self._credentials = GoogleCredentials.get_application_default()
+    self._tpu_service = discovery.build(
+        'tpu', 'v1', credentials=self._credentials, cache_discovery=False)
+    self._compute_service = discovery.build(
+        'compute', 'v1', credentials=self._credentials, cache_discovery=False)
+
+    if project is None:
+      self._project = self._get_instance_metadata('project/project-id')
+    if zone is None:
+      zone_path = self._get_instance_metadata('instance/zone')
+      self._zone = zone_path.split('/')[-1]
+    self._vm_master = self._get_instance_metadata('instance/name')
+
+  def _get_instance_group(self):
+    """Gets the instance group that the current VM belongs to."""
+    resp = self._compute_service.instances().get(
+        project=self._project,
+        zone=self._zone,
+        instance=self._vm_master,
+        fields='metadata').execute()
+
+    if 'metadata' in resp and 'items' in resp['metadata']:
+      for item in resp['metadata']['items']:
+        if (item['key'] == 'created-by' and
+            'instanceGroupManagers' in item['value']):
+          return item['value'].split('/')[-1]
+
+    raise RuntimeError(('A vm list must be passed to ClusterResolver '
+                        'if not using an instance group'))
+
+  def _get_member_instance_names(self, instance_group):
+    """Gets all the instance names that belong to the given instance group."""
+    resp = self._compute_service.instanceGroups().listInstances(
+        project=self._project, zone=self._zone,
+        instanceGroup=instance_group).execute()
+
+    instances = []
+    if 'items' in resp:
+      for item in resp['items']:
+        if 'instance' not in item or 'status' not in item:
+          continue
+        instance_path = item['instance']
+        instances.append(instance_path.split('/')[-1])
+
+    return instances
+
+  def _get_client_workers(self):
+    """Gets client workers.
+
+    The instance group that the current VM belongs to is picked up from
+    the GCE instance metadata set of the VM. If a list of VMs was used for
+    initializing cluster resolver, we use that instead.
+
+    Returns:
+      A list of ClientWorker.
+
+    Raises:
+      RuntimeError: If the red VM cluster is not healthy.
+    """
+    if self._vms is None:
+      # Using an instance group
+      instance_group = self._get_instance_group()
+      self._vms = self._get_member_instance_names(instance_group)
```

Comment:
Is it useful to log a warning or throw an error is self._vms is still empty after running this?

Reply:
This would mean that the GCE InstanceGroup product is broken in that case, but done. I have added this check and corresponding test. Thanks.